odl/contrib/solvers/spdhg/misc.py

```diff
@@ -0,0 +1,857 @@
+
+                    z_ /= sigmaSqrt
+                    grad = sigmaSqrt * self.grad
+                    grad.norm = sigmaSqrt * self.grad.norm
+                    niter = opts['niter']
+                    alpha = self.alpha
+                    out[:] = FGP_dual(p, z_, alpha, niter, grad, proj_C,
+                                      proj_P, tol=opts['tol'])
+
+                    out *= sigmaSqrt
+
+                    return out
+
+                else:
+                    raise NotImplementedError('Not yet implemented')
+
+            return TV_prox
+
+
+def FGP_dual(p, data, alpha, n_iter, grad, proj_C, proj_P, tol=None, **kwargs):
+    """ Computes a saddle point with the PDHG method.
+
+    Parameters
+    ----------
+    p : np.array
+        dual initial variable
+    data : np.array
+        noisy data / proximal point
+    alpha : float
+        regularization parameter
+    n_iter : int
+        number of iterations
+    grad : gradient class
+        class that supports grad(x), grad.adjoint(x), grad.norm
+    proj_C : function
+        projection onto the constraint set of the primal variable,
+        e.g. non-negativity
+    proj_P : function
+        projection onto the constraint set of the dual variable,
+        e.g. norm <= 1
+    tol : float (optional)
+        nonnegative parameter that gives the tolerance for convergence. If set
+        None, then the algorithm will run for a fixed number of iterations
+
+    Other Parameters
+    ----------------
+    callback : callable, optional
+        Function called with the current iterate after each iteration.
+    """
+
+    # Callback object
+    callback = kwargs.pop('callback', None)
+    if callback is not None and not callable(callback):
+        raise TypeError('`callback` {} is not callable'
+                        ''.format(callback))
+
+    factr = 1 / (grad.norm**2 * alpha)
+
+    q = p.copy()
+    x = data.space.zero()
+
+    t = 1.
+
+    if tol is None:
+        def convergenceEval(p1, p2):
+            return False
+    else:
+        def convergenceEval(p1, p2):
+            return (p1 - p2).norm() / p1.norm() < tol
+
+    pnew = p.copy()
+
+    if callback is not None:
+        callback(p)
+
+    for k in range(n_iter):
+        t0 = t
+        grad.adjoint(q, out=x)
+        proj_C(data - alpha * x, out=x)
+        grad(x, out=pnew)
+        pnew *= factr
+        pnew += q
+        proj_P(pnew, out=pnew)
+
+        converged = convergenceEval(p, pnew)
+
+        if not converged:
+            # update step size
+            t = (1 + np.sqrt(1 + 4*t0**2))/2.
+
+            # calculate next iterate
+            q[:] = pnew + (t0 - 1)/t * (pnew - p)
+
+        p[:] = pnew
+
+        if converged:
+            t = None
+            break
+
+        if callback is not None:
+            callback(p)
+
+    # get current image estimate
+    x = proj_C(data - alpha * grad.adjoint(p))
+
+    return x
+
+
+class Blur(odl.Operator):
```

Comment:
`Blur2d` since it's only implemented in 2D.

Reply:
done